subsys/net/lib/nrf_cloud/src/nrf_cloud_logs.c

```diff
@@ -0,0 +1,257 @@
+ * to using the full logger backend above.
+ */
+#if defined(CONFIG_NRF_CLOUD_MQTT)
+int nrf_cloud_log_send(int log_level, const char *fmt, ...)
+{
+	va_list ap;
+	int err;
+
+	/* Do we want any level filtering in this simple logging configuration? */
+	if (log_level > nrf_cloud_log_level) {
+		LOG_DBG("Filtering out msg; level %d > %d", log_level, nrf_cloud_log_level);
+		return 0;
+	}
+
+#if defined(CONFIG_NRF_CLOUD_LOG_BACKEND)
+	/* If cloud logging backend is enabled, send it through the main logging system. */
+	va_start(ap, fmt);
+	log_inject(log_level, fmt, ap);
+	va_end(ap);
+	err = 0;
+#else
+	struct nrf_cloud_log_context context = {0};
+	char buf[CONFIG_NRF_CLOUD_LOG_BUF_SIZE];
+	struct nrf_cloud_tx_data output = {
+		.qos = MQTT_QOS_0_AT_MOST_ONCE,
+		.topic_type = NRF_CLOUD_TOPIC_MESSAGE
+	};
+
+	/* Send it directly to the cloud. */
+	va_start(ap, fmt);
+	err = log_format(&context, buf, &output, NULL, NULL, log_level, fmt, ap);
+	va_end(ap);
+	if (err) {
+		return err;
+	}
+	err = nrf_cloud_send(&output);
+	if (err) {
+		LOG_ERR("Error sending log:%d", err);
+	}
+#endif /* CONFIG_NRF_CLOUD_LOG_BACKEND */
+
+	return err;
+}
+#endif /* CONFIG_NRF_CLOUD_MQTT */
+
+#if defined(CONFIG_NRF_CLOUD_REST)
+int nrf_cloud_rest_log_send(struct nrf_cloud_rest_context *ctx, const char *dev_id,
+			     int log_level, const char *fmt, ...)
+{
+	__ASSERT_NO_MSG(ctx != NULL);
+	__ASSERT_NO_MSG(dev_id != NULL);
+	va_list ap;
+	int err;
+
+	/* Do we want any level filtering in this simple logging configuration? */
+	if (log_level > nrf_cloud_log_level) {
+		LOG_DBG("Filtering out msg; level %d > %d", log_level, nrf_cloud_log_level);
+		return 0;
+	}
+
+#if defined(CONFIG_NRF_CLOUD_LOG_BACKEND)
+	/* Cloud logging backend is enabled, so send it through the main logging system. */
+	va_start(ap, fmt);
+	log_inject(log_level, fmt, ap);
+	va_end(ap);
+	err = 0;
+#else
+	struct nrf_cloud_log_context context = {0};
+	char buf[CONFIG_NRF_CLOUD_LOG_BUF_SIZE];
+	struct nrf_cloud_tx_data output = {
+		.qos = MQTT_QOS_0_AT_MOST_ONCE,
+		.topic_type = NRF_CLOUD_TOPIC_MESSAGE
+	};
+
+	/* Send it directly to the cloud. */
+	va_start(ap, fmt);
+	err = log_format(&context, buf, &output, NULL, NULL, log_level, fmt, ap);
+	va_end(ap);
+	if (err) {
+		return err;
+	}
+
+	err = k_sem_take(&ncl_active, K_MSEC(REST_WAIT_MS));
+	if (err < 0) {
+		LOG_INF("Failed to take semaphore");
+		return err; /* Caller should try again; busy or other error */
+	}
+	err = nrf_cloud_rest_send_device_message(ctx, dev_id,
+						 output.data.ptr, false, NULL);
+	k_sem_give(&ncl_active);
+
+	if (err) {
+		LOG_ERR("Error sending log:%d", err);
+	}
+#endif /* CONFIG_NRF_CLOUD_LOG_BACKEND */
+
+	return err;
+}
+#endif /* CONFIG_NRF_CLOUD_REST */
+
+void nrf_cloud_log_control_set(int level)
+{
+	if (nrf_cloud_log_level != level) {
+		LOG_DBG("Changing log level from:%d to:%d", nrf_cloud_log_level, level);
+		nrf_cloud_log_level = level;
+	}
+	nrf_cloud_log_enable(level != LOG_LEVEL_NONE);
```

Comment:
The function implicitly enables/disables logging when logging level is changed. Just thinking that this might not be clear when calling the API and whether it would be better to let user call `nrf_cloud_log_enable()` separately when needed? Does it matter if logging remains enabled when log level is set to NONE?

Reply:
They each serve different purposes but are interrelated.

- `nrf_cloud_log_control_set()` is normally only called when a shadow delta occurs (usually due to the user changing the log level in the Log card in the portal).  The portal and the shadow only have a log level value, not also an enable flag.  We use LOG_LEVEL_NONE from the cloud side to disable cloud logging.
- `nrf_cloud_log_enable()` is called from `nrf_cloud_log_control_set()` whenever such a shadow delta is received in order to enable or disable the backend as needed
- separately, `nrf_cloud_log_enable()` needs to be called to turn on the logging backend when the network is available; it is purposely not automatically enabled on startup but only when a REST-based application is ready for network use, or when an MQTT-based application is first connected (when the shadow is received)

Reply:
Ok, I guess that's the reason function name has the word "control" in it instead of "level". Ideally, a function should do what the name says. The word "control" is vague, so at least the function description should describe what the function does. Now it only mentions the the function sets the logging level, so enabling/disabling of logging is hidden logic.

This is a public API, so we should consider usability from that point of view, not how it's used by a specific piece of code. `nrf_cloud_log_level_set()` and `nrf_cloud_log_enable()` would be self-explanatory, but that's just my opinion.

Reply:
Control in this case came from the fact that we are storing the alertEn flag and the logLvl value for the alerts and logs services in a section of the shadow called control.  The similar alerts functions are `nrf_cloud_alert_control_set()` and `nrf_cloud_alert_control_get()`.

I could make these 4 functions internal and not list them in the public include files.

Reply:
But then there would be no public API for setting the log level, just wondering if it should be available?

Reply:
If there is no known use case for setting the logging level and enabling/disabling logging at runtime through public API, I'd leave them internal. We can make them public if needed but generally logging is configured at compile time.
I understand there is a need to change logging level runtime through cloud, but as said, that doesn't require public API for it.

Reply:
I went ahead and made a public function to set the log level.  I also improved the header docs to explain better what the nrf_cloud_log_control_set() function does.  I am considering removing this but have not decided yet.